packages/langium/src/documents/document-builder.ts

```diff
@@ -29,17 +28,19 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
         this.documentValidator = services.validation.DocumentValidator;
     }
 
-    build(document: LangiumDocument): void {
+    build(document: LangiumDocument, diagnostics?: Diagnostic[]): void {
```

Comment:
I feel we should find a cleaner solution than passing a mutable array. What about returning an object like this?
```typescript
return {
    get diagnostics() {
         // If not done before, compute the diagnostics, otherwise return the previously computed diagnostics.
    }
};
```

Reply:
We could add even more data to the result object:
```typescript
export type BuildResult = {
     readonly parseResult: ParseResult<AstNode>
     readonly diagnostics: Diagnostic[]
}
```
With this, we can omit the `undefined` check when accessing the parse result from the document.

Reply:
Yeah, I like this way better.